Blake2b F precompile

mhluongo opened the pull request:
Adapt https://github.com/ethereum/EIPs/issues/152 into the proper format.

Confirming a couple details before I open this against upstream. cc @pdyraga

mhluongo committed:
Brought issue #152 into the repo as a draft EIP.

Thanks @tjade273!

mhluongo committed:
Make the draft EIP consistent with the template

Also added myself as an author

mhluongo committed:
Break backwards compatibility into its own section

mhluongo committed:
Added notes about the in-progress implementation

Should have a working geth precompile and initial benchmarks shortly

mhluongo committed:
Specify EIP-2046 as a requirement

While 2046's cheaper precompile contract calls aren't a requirement for
this EIP's implementation, shipping this precompile without EIP-2046
would make the F function expensive for some of the motivating usecases.

pdyraga commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,76 @@
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+Adds a precompile at address `0x0d` which accepts [ABI encoded](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI) arguments corresponding to the function signature
```
I wonder why `0x0d`, all currently available precompiles are between `[0x01, 0x08]` so the next available slot is `0x09` 🤔

mhluongo replied:
Legacy from the original EIP- other pending EIPs have claimed past `0x08`

mhluongo replied:
We'll leave it to an EIP editor to change that

pdyraga commented on EIPS/eip-draft-blake2b-f-precompile.md:
````diff
@@ -0,0 +1,76 @@
+F(bytes32[2] h, bytes32[4] m, uint t , bool f, uint rounds) returns (bytes32[2] h_new);
+```
+
+where `h`, `m`, `t` and `f` are the current state, the new message, the byte counter and a finalization flag, as defined in [RFC 7693](https://tools.ietf.org/html/rfc7693), and `rounds` is the number of rounds of mixing to perform (BLAKE2b uses 12, BLAKE2s uses 10). `h_new` is the updated state of the hash.
````
Should we allow an arbitrary number of rounds? Our library supports `[0,12]` rounds and we have confirmed this is all Zcash needs.

mhluongo replied:
So [the RFC](https://tools.ietf.org/html/rfc7693) suggests 12 is the most we need- however! I suggest we allow more rounds using the same extension trick blake2b uses. I'll comment in the implementation

pdyraga replied:
Updated the implementation, we now support any number of rounds.

pdyraga commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,76 @@
+## Implementation
+<!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->
+
+Implementations are in progress, and can be followed along in our [Golang Blake2 library fork](https://github.com/keep-network/blake2-f) as well as our fork of [go-ethereum](https://github.com/keep-network/go-ethereum).
```
This is only `F` function implementation extracted from Golang Blake2 lib, should we make it clear here? 
(Original Golang Blake2 lib does not export `F` out)

pdyraga commented on EIPS/eip-draft-blake2b-f-precompile.md:
````diff
@@ -0,0 +1,76 @@
+Adds a precompile at address `0x0d` which accepts [ABI encoded](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI) arguments corresponding to the function signature
+
+```
+F(bytes32[2] h, bytes32[4] m, uint t , bool f, uint rounds) returns (bytes32[2] h_new);
````
Our `F` Go implementation has the following signature:
```
F(h *[8]uint64, blocks []byte, t *[2]uint64, f uint64, rounds uint8)
```
- `h [8]uint64` = `8 * 64 bits = 512 bits = 2 * 32 * 8 bits = 2 * 32 bytes` = `bytes32[2] h`
- `blocks []byte` can fit `bytes32[4] m` but worth to confirm with ZCash if it's enough for all their use cases 
- `t [2]uint64` is not compatible with `uint t`
- `f uint64` is not compatible with `bool f`
- `rounds uint8` is generally compatible with `uint rounds` because we currently allow for `[0,12]` rounds.

What's interesting, the Go implementation from Tjaden [has a similar function signature](https://github.com/tjade273/go-ethereum/blob/39433efa7aff22f426946aab57c44dce88db6fb7/crypto/blake2b/blake2.go#L283) to ours:
```
func BlakeCompress(h [8]uint64, m []uint8, t [2]uint64, f [2]uint64) [8]uint64
```   
I am not sure what is the reason for `t` and `f` to be just `uint/bool` here.

pdyraga replied:
After rethinking it:
- ZCash uses both BLAKE2s and BLAKE2b. For 2s, `h` could be represented as `[8]uint32`and for 2b, `h` could be represented as `[8]uint64` on Go side. We are implementing a generic function `F` so it's indeed better to pack the state vector into `bytes32[2]` on Solidity side and use `[8]uint64` on Go side. If needed, two Solidity functions delegating to generic `F` can be exposed later: one with `bytes8[8] h` for 2b and one with `bytes4[8] h` for 2s.
- block size is `128` so `bytes32[4]` fits nicely here on Solidity side. We can even refactor Go `F` implementation to accept `blocks [BlockSize]byte` instead of just `blocks []byte`
- According to the RFC, `f` can be boolean:
    ```
     IF f = TRUE THEN              // last block flag?
       v[14] := v[14] ^ 0xFF..FF   // Invert all bits.
     END IF.
    ```
    I will refactor Go `F` implementation to accept `bool` instead of `uint64` to avoid confusion.
- I still can't agree about `t [2]uint64`. The RFC says `t` is a `2w-bit offset counter`. For 32, we should have `[2]uint32` in Go and for 64 we should have `[2]uint64` in Go. Since `uint` in Solidity defaults to `uint256` we have more than required there. Why not `uint128` in Solidity?

mhluongo replied:
Is there actually a benefit from using a `uint128` in Solidity? I believe `uint256` is a single slot on the EVM- so we should always pack as tightly as we can into that base unit.

mhluongo replied:
Picking the packing issue up in a more public EIP PR

pdyraga replied:
> Is there actually a benefit from using a uint128 in Solidity? I believe uint256 is a single slot on the EVM- so we should always pack as tightly as we can into that base unit.

Very good point but I think it matters only when we store data on the chain. Here we just pass parameters so it should be all in-memory.

pdyraga replied:
I am going to resolve this conversation since we decided not to ABI-encode the parameters.

pdyraga commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,76 @@
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+Adds a precompile at address `0x0d` which accepts [ABI encoded](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI) arguments corresponding to the function signature
```
Per our discussion, we should update the part about ABI encoded parameters.

carterpy replied:
@pdyraga, I've been playing with your f function in go and it's great and you've done a fantastic job here - massive props. I have two questions but please do not answer the stupid ones: 
1) In F.go, should f return h? 
2) Should the package be the same name as the folder it's in?

carterpy replied:
Another silly question: RFC7693 has 12 rounds but we're doing 10. I trust you guys but why 10 instead of 12?

mhluongo replied:
I recommend we move implementation discussion to the [blake2-f library's first PR](https://github.com/keep-network/blake2-f/pull/1)- WDYT?

pdyraga replied:
@carterpy Can you please repost your questions in the PR referenced by @mhluongo? I can happily answer them there.

PlainsWraith replied:
@pdyraga, questions removed because they were actually silly :$

pdyraga replied:
Here is an alternative proposition about how to call the `F` function.

Instead of ABI-encoded parameters (no precompile currently do that) we can use a standard `staticcall`. 

I'd propose the following structure:
```
[ bytes32[0] | bytes32[1] ][ bytes32[2] | bytes32[3] | bytes32[4] | bytes32[5] ]
[      64 bytes for h     ][                    128 bytes for m                ]

[             bytes32[6]             ][             bytes32[7]             ]
[ 24 bytes padding | 8 bytes for t_0 ][ 24 bytes padding | 8 bytes for t_1 ]

[             bytes32[8]          ][              bytes32[9]              ]
[ 31 bytes padding | 1 byte for f ][ 28 bytes padding | 4 bytes for rounds]
```
where:
- `h` - state vector
- `m` - message block vector
- `t_0, t_1` - offset counter
- `f` - final block indicator flag
- `rounds` - number of rounds

This way we pack data in equal 32-byte (256-bit) slots, we are compatible with the rest of precompile calls and we do not require any bitwise juggling from callers.

For convenience, one may expose a wrapping function, similar to `F`:
```
    function F(bytes32[2] memory h, bytes32[4] memory m, uint64[2] memory t, bool f, uint32 rounds) public view returns (bytes32[2] memory) {
      bytes32[2] memory output;

      bytes32 _t0 = bytes32(uint256(t[0]));
      bytes32 _t1 = bytes32(uint256(t[1]));

      bytes32 _f;
      if (f) {
        _f = hex"0000000000000000000000000000000000000000000000000000000000000001";
      }

      bytes32 _rounds = bytes32(uint256(rounds));

      bytes32[10] memory args = [ h[0], h[1], m[0], m[1], m[2], m[3], _t0, _t1, _f, _rounds ];

      assembly {
            if iszero(staticcall(not(0), 0x09, args, 0x140, output, 0x40)) {
                revert(0, 0)
            }
        }
        return output;
    }

    function callF() public view returns (bytes32[2] memory) {
      bytes32[2] memory h;
      h[0] = hex"6a09e627f3bcc909bb67ae8484caa73b3c6ef372fe94b82ba54ff53a5f1d36f1";
      h[1] = hex"510e527fade682d19b05688c2b3e6c1f1f83d9abfb41bd6b5be0cd19137e2179";

      bytes32[4] memory m;
      m[0] = hex"278400340e6b05c5752592c52c4f121292eafc51cc01a997b4aed13409298fad";
      m[1] = hex"0d99ccc8f76453d9b3661e5c4d325d7751147db17489046d1682d50eefa4a1da";
      m[2] = hex"0000000000000000000000000000000000000000000000000000000000000000";
      m[3] = hex"0000000000000000000000000000000000000000000000000000000000000000";

      uint64[2] memory t;
      t[0] = 18446744073709551552;
      t[1] = 18446744073709551615;

      bool f = true;

      uint32 rounds = 12;

      return F(h, m, t, f, rounds);
    }
```

pdyraga replied:
(this looks terrible at the first sight, but this is similar to how other precompiles are called with assembly 🤷‍♀ - they are all equally terrible).

pdyraga commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,76 @@
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
+
+## Specification
```
Discussing gas price based on [benchmarks from the implementation](https://github.com/keep-network/go-ethereum/pull/4#discussion_r295574236) and using the same approach as we did for [EIP-1108](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1108.md)...

Let's assume `ecrecover` is perfectly priced. It costs `3000` gas and takes `180.3` microseconds on my Mac to execute. It's about `16.66` gas per microsecond of `ecrecover` work. 

With this in mind, `F` took `402` nanoseconds on average, it is equal to `0.402` microseconds so the invocation (assuming the same input parameters) should cost `0.402 * 16.66` = `6.7` gas.

Those benchmarks were done with 12 rounds, so it's `0.56` per round.

In the original EIP there was mentioned a price of `37` gas per round. 

We have it either heavily optimized or I am too tired to see where is the mistake in my calculations...

mhluongo committed:
Don't use ABI encoding for precompile

Replace the existing ABI encoding interface to the BLAKE2b `F`
precompile with a loosely pack struct that's `staticcall`-friendly.

H/t to @pdyraga for putting together the interface!

prestwich commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,142 @@
+
+The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors.
+
+Interoperability with Zcash could enable contracts like trustless atomic swaps between the chains, which could provide a much needed aspect of privacy to the very public Ethereum blockchain.
```
Might also mention Sia and Handshake?

mhluongo committed:
Add @pdyraga to the EIP authors.

mhluongo committed:
Remove less relevant EIP rationale

Let's not relitigate precompiles, WASM, etc in thie EIP :)

prestwich commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,142 @@
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+We propose adding a precompiled contract at address `0x09` wrapping the [BLAkE2b `F` compression function](https://tools.ietf.org/html/rfc7693#section-3.2). Rather than the ABI-encoded call structure proposed in [#152](https://github.com/ethereum/EIPs/issues/152), this design better matches other precompiles and works well with a standard `staticcall`.
```
example of other

prestwich commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,142 @@
+## Rationale
+<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised  ...
+
+The most frequent concern with EIPs of this type is that the addition of specific functions at the protocol level is an infringement on Ethereum's "featureless" design. It is true that a more elegant solution to the issue is to simply improve the scalability characteristics of the network so as to make calculating functions requiring millions of gas practical for everyday use. In the meantime, ho ...
```
don't say I

prestwich commented on EIPS/eip-draft-blake2b-f-precompile.md:
```diff
@@ -0,0 +1,142 @@
+## Backwards Compatibility
+<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
+
+There is very little risk of breaking backwards-compatibility with this EIP, the sole issue being if someone were to build a contract relying on the address at `0x0d` being empty. The likelihood of this is low, and should specific instances arise, the address could be chosen to be any arbitrary value with negligible risk of collision.
```
`0x09` or `0x0d`?

mhluongo committed:
Use 0x09 as the precompile address

If a conflicting EIP is moving forward the EIP editor can assign a new
address

mhluongo commented:
@prestwich thanks fore the comments- will address in a more public EIP

mhluongo commented:
Closing in favor of an upstream PR https://github.com/ethereum/EIPs/pull/2129
